### PR TITLE
[LMS-337] Using course alias for SIS section identifier

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py [hunk 1]
```diff
@@ -20,6 +20,178 @@
...
+
+    Returns
+    -------
+    List[Dict[str, str]]
+        a list of Google Classroom CourseAliases for the given course id,
+            see https://developers.google.com/classroom/reference/rest/v1/courses.aliases
+    """
+
+    if resource is None:
+        return []
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:44]: Line 44 is not covered by unit tests. Probably not a big deal. How much effort & value would come from adding a test?

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py [hunk 2]
```diff
@@ -20,6 +20,178 @@
...
+    List[str]
+        a list of alias names (scope omitted) filtered by the given scope
+    """
+    result: List[str] = []
+    try:
+        result = [
+            alias.split(":")[1] for alias in aliases if alias.split(":")[0] == scope
+        ]
+    except (ValueError, IndexError):
+        pass
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:77]: Any value to having a unit tests for the `except` case? I think the main thing would be insuring that the calling function handles the empty list.

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py [hunk 3]
```diff
@@ -20,6 +20,178 @@
...
+    if edfilms_project_alias != "":
+        return edfilms_project_alias
+
+    if len(domain_alias_names) > 0:
+        return domain_alias_names[0]
+
+    if len(project_alias_names) > 0:
+        return project_alias_names[0]
+
+    return ""
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:169]: Another edge case without a test. Are we adequately testing what happens if no aliases at all have been setup?

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py [hunk 4]
```diff
@@ -20,6 +20,178 @@
...
+        try:
+            _, alias_name = aliases[0].split(":")
+            return alias_name
+        except ValueError:
+            return ""
+
+    domain_alias_names: List[str] = _alias_names_for_scope(aliases, "d")
+    edfilms_domain_alias = _find_first_with_edfilms_prefix(domain_alias_names)
+    if edfilms_domain_alias != "":
+        return edfilms_domain_alias
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:156]: Unless I'm overlooking something, the output file will have `SectionIdentifier = EdFiLMSxyz` in it - which won't match the SIS. Should we strip off `EdFiLMS`? And if so, should we use any delimiter, e.g. "EdFiLMS.xyz"? No one ever sees the value, so I'm not sure that it is all that useful.

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py (hunk 3 above)
Comment (stephenfuqua) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:169]: I have uploaded some aliases and the project aliases aren't coming through. For Spanish III I loaded "p:XYZ" and for English I I uploaded ["p:a1b2", "p:EdFiLMS.a1b2_"]. In both cases, the extractor output a blank value for the `SISSectionIdentifier`.

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py (hunk 1 above)
Comment (bradbanister) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:44]: I think resource being Optional is outdated. That check is largely to keep mypy happy. Removed Optional and those checks, everything looks good.

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py (hunk 2 above)
Comment (bradbanister) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:77]: Added test for malformed aliases.

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py (hunk 4 above)
Comment (bradbanister) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:156]: Removing prefix.

File: src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py (hunk 3 above)
Comment (bradbanister) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:169]: Line 169 handles multiple unknown scopes without a known scope. Added test.

Comment (bradbanister) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:169]: The GC API isn't returning any aliases for Spanish III. Maybe because the course is status ARCHIVED?

Comment (stephenfuqua) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:169]: I suspect a permission problem, because I'm able to access that from the [API tester](https://developers.google.com/classroom/reference/rest/v1/courses.aliases/list)
image

Comment (bradbanister) [reply src/google-classroom-extractor/edfi_google_classroom_extractor/api/courses.py:153]: Removed project scope support, doesn't look relevant.

File: src/google-classroom-extractor/README.md [hunk 5]
```diff
@@ -177,6 +177,22 @@ for more information."`
...
+
+## Course Aliases and SIS Section Identifiers
+
+Course Aliases in Google Classroom are expected to be used to provide a mapping
+from a Google Classroom Course to a SIS section using a SIS section identifer.
+To enable this behavior, create a domain-scoped Course Alias in Google Classroom
+for each Course with the prefix "EdFiLMS." followed by the SIS section identifier.
+For example, the domain-scoped Course Alias "d:EdFiLMS.ALG-123" would be used for
+SIS section identifier "ALG-123". Only the first Course Alias found with the
+"EdFiLMS." prefix will be used.
```
Comment (stephenfuqua) [inline src/google-classroom-extractor/README.md:189]: 👍

**Review by stephenfuqua — APPROVED** [review]
